### ISSUE TITLE
how to tune hyp when using focal loss?

### ISSUE THREAD

**Issue by Xiaocong6** [ultralytics/yolov5#2864]
## ❔Question
hi~ when i used the focal loss(fl_gamma=1.5), the confidence of the prediction becomes very low(<0.5).
Do I need to adjust the weights of the three losses? （hyp['box'] hyp['cls'] hyp['obj']）

**Comment by glenn-jocher** [comment]
@Xiaocong6 low confidence on predictions is somewhat on purpose in YOLOv5. This is because the objectness outputs are being trained to match the actual IoU (of predicted box vs labelled box) rather than being trained to match 1.0. You can set this line to 1.0 to increase confidence (especially early on or for short trainings), though this may also slightly reduce mAP.
```python
tobj[b, a, gj, gi] = 1.0
```

https://github.com/ultralytics/yolov5/blob/1df8c6c963d31ce84895101a70e45e0afdcb0bc2/utils/loss.py#L135-L137

**Comment by Xiaocong6** [comment]
Thanks!
You see the photo. It is the printed losses(column 2-4) in different epochs(column 1). The difference between the three losses is a bit large. Do you think I need to adjust the weights of these three losses to make their values close?
![image](https://user-images.githubusercontent.com/44869509/115483412-4d866180-a283-11eb-922a-1ecfaab48a8f.png)

**Comment by glenn-jocher** [comment]
@Xiaocong6 they seem pretty similar, but if you'd like to modify them you're free to adjust the loss gain hyps here:
https://github.com/ultralytics/yolov5/blob/5f7d39fede4de8af98472bd009c63c3a86568e2d/data/hyp.scratch.yaml#L13-L17

**Comment by Xiaocong6** [comment]
Thank you very much!

**Comment by julian-douglas** [comment]
Hi Glenn, 

What exactly do 'box', 'cls' and 'obj' do? I have been searching for a while and I can't put a finger on what exactly they do. What effect would increasing or decreasing these have?

**Comment by glenn-jocher** [comment]
@julian-douglas these are loss gains. Comments in hyp:
https://github.com/ultralytics/yolov5/blob/1d65e8194d183f4e1537cf64a7292f8ab57d1d55/data/hyps/hyp.scratch.yaml#L13-L17

Used in loss.py

**Comment by julian-douglas** [comment]
Hi Glenn,

Thanks for that response! Except I don’t actually know what ‘loss gain’ means? What role does it play on the loss functions? I’ve tried googling but i can’t find anything of use.

**Comment by glenn-jocher** [comment]
@julian-douglas see loss.py:
https://github.com/ultralytics/yolov5/blob/1d65e8194d183f4e1537cf64a7292f8ab57d1d55/utils/loss.py#L163-L169

**Comment by Stephen-Hao** [comment]
@glenn-jocher Hello Glenn,
Thank you very much!  
When the training set has less labeled data, more than 1 : 100 (Positive sample with labels : negative samples without labels), the confidence of the prediction becomes very low(<0.01) too.   
I can use focal loss at the same time and modify the "cls_pw" and "obj_pw" a lot, such as greater than 4.0, which method is more effective for the imbalance of foreground, background ?  
Thanks a lot, Best wishes! 
![image](https://user-images.githubusercontent.com/35420905/192703740-4876eeb4-49ce-408a-beda-9177bb30279d.png)

**Comment by Stephen-Hao** [comment]
> When the training set has less labeled data, more than 1 : 100 (Positive sample with labels : negative samples without labels), the confidence of the prediction becomes very low(<0.01) too.
> I can use focal loss at the same time and modify the "cls_pw" and "obj_pw" a lot, such as greater than 4.0, which method is more effective for the imbalance of foreground, background ?

When the training set has less labeled data, more than 1 : 100 (Positive sample with labels : negative samples without labels), the confidence of the prediction becomes very low(<0.01) too.
I can use focal loss at the same time and modify the "cls_pw" and "obj_pw" a lot, such as greater than 4.0, which method is more effective for the imbalance of foreground, background ?

**Comment by glenn-jocher** [comment]
@Stephen-Hao yes you can set the focal loss here, but be very careful as this easily leads to many FPs. You can also experiment with setting class weights in the loss functions if you are referring to class imbalance issues.
